sql/core/src/main/scala/org/apache/spark/sql/jdbc/JdbcDialects.scala

```diff
@@ -240,3 +242,21 @@ case object DB2Dialect extends JdbcDialect {
     case _ => None
   }
 }
+
+/**
+ * :: DeveloperApi ::
+ * Default SQL Server dialect, mapping the datetimeoffset types to a String on read.
+ */
+@DeveloperApi
+case object SqlServerDialect extends JdbcDialect {
+
+  override def canHandle(url: String): Boolean = url.startsWith("jdbc:sqlserver")
+
+  override def getCatalystType(
+    sqlType: Int, typeName: String, size: Int, md: MetadataBuilder): Option[DataType] = {
+        if (typeName.contains("datetimeoffset")) {
+          // String is used by SQL Server for datetimeoffset types in legacy clients
```

Comment:
The "legacy clients" comes from this [Microsoft Technet page](https://technet.microsoft.com/en-us/library/ms180878%28v=sql.105%29.aspx#BackwardCompatibilityforDownlevelClients) I think they mean "any client which isn't SQL Server specific".

For JDBC they say use a type mapping of "Java.sql.String"